solidity/contracts/BancorUpgradeConverter.sol

```diff
@@ -0,0 +1,236 @@
+        copyConvertionFee(_fromConverter, toConverter);
+        copyQuickBuyPath(_fromConverter, toConverter);
+        transferConnectorsBalances(_fromConverter, toConverter, formerVersions);
+        _fromConverter.transferTokenOwnership(toConverter);
+        toConverter.acceptTokenOwnership();
+        _fromConverter.transferOwnership(msg.sender);
+        toConverter.transferOwnership(msg.sender);
+        toConverter.transferManagement(msg.sender);
+        ConverterCreated(address(_fromConverter), address(toConverter));
+    }
+
+    /**
+        @dev first step when upgrading a converter is to transfer the ownership to the upgrade converter
+        contract instance. since transferring ownership has two steps, the instance must to accept the
+        ownership before it can proceed the upgrading process.
+        notice that even that the ownership transfered to the contract, only the current owner can 
+        execute the upgrade process.
+        when the process is finish, the function fired an event.
+
+        @param _fromConverter       converter to accept its ownership request
+
+        @return the new converter
+    */
+    function acceptConverterOwnership(IBancorConverter _fromConverter) private {
+        _fromConverter.acceptOwnership();
+        ConverterOwned(_fromConverter, this);
+    }
+
+    /**
+        @dev create a new converter with same token and the same max conversion fee and the same 
+        extensions address as set in the given converter.
+        the new converter has no connectors.
+
+        @param _fromConverter       converter to read from
+
+        @return the new converter
+    */
+    function createConverter(IBancorConverter _fromConverter) private returns(IBancorConverter) {
+        IERC20Token emptyConnector = IERC20Token(address(0));
+        ISmartToken token = _fromConverter.token();
+        IBancorConverterExtensions extensions = _fromConverter.extensions();
+        uint32 maxConversionFee = _fromConverter.maxConversionFee();
+        BancorConverter converter = new BancorConverter(
+            token,
+            extensions,
+            maxConversionFee,
+            emptyConnector,
+            0
+        );
+        return IBancorConverter(converter);
+    }
+
+    /**
+        @dev copies the connectors from the current converter to the new one
+
+        @param _fromConverter       current converter to copy from
+        @param _toConverter         new converter to add connectors
+        @param _formerVersions      true if the converter version is under 0.5
+    */
+    function copyConnectors(IBancorConverter _fromConverter, IBancorConverter _toConverter, bool _formerVersions)
+        private
+    {
+        uint256 virtualBalance;
+        uint32 weight;
+        bool isVirtualBalanceEnabled;
+        bool isPurchaseEnabled;
+        bool isSet;
+        for (uint16 i = 1; i < _fromConverter.convertibleTokenCount(); i++) {
```

Comment:
this will not work for an unlimited number of connectors, should document it.

also, convertibleTokenCount will also return the smart token itself. Should use connectorTokenCount instead.

Reply:
changed to connectorTokenCount